### PR TITLE
[telemetrySettting] Create sampled Logger

### PR DESCRIPTION
**Pull request by antonjim-te**
**Issue**: 

In one of the use cases which we are working on, we have identified that `sampledLogger` mechanism used in the Opentelemetry exporters (`otlpexporter` and `otlphttpexporter`), could become a memory bottleneck, when the number of exports in the same pipeline is high (we are talking about 100s of instances of exporters running on the same machine).

The sampledLogger is created inside the exporterhelper module and the sampling is always activated unless the collector is in debug mode:

- https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L88

```
func newQueuedRetrySender(id component.ID, signal component.DataType, qCfg QueueSettings, rCfg RetrySettings, reqUnmarshaler internal.RequestUnmarshaler, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
	retryStopCh := make(chan struct{})
	sampledLogger := createSampledLogger(logger)
```
- https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L269
```
func createSampledLogger(logger *zap.Logger) *zap.Logger {
	if logger.Core().Enabled(zapcore.DebugLevel) {
		// Debugging is enabled. Don't do any sampling.
		return logger
	}

	// Create a logger that samples all messages to 1 per 10 seconds initially,
	// and 1/100 of messages after that.
	opts := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
		return zapcore.NewSamplerWithOptions(
			core,
			10*time.Second,
			1,
			100,
		)
	})
	return logger.WithOptions(opts)
}
```
**Investigation:**

The high memory usage has been discovered using https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/pprofextension

- In the first scenario, default configuration with the `sampledLogger` active, based on the profiled data we could observe that sampling mechanism was responsible for allocating aprox. 75% of the memory of the running collector:
image

- In the second scenario, with the same number of running exporters, but with the `sampledLogger` mechanism disabled (using `debug` level configuration check configuration below), we were able to observe a much smaller memory footprint:
```
    service:
      telemetry:
        logs:
          level: debug
```
image

Given that the logging sampling mechanism could become a bottleneck in certain scenarios (our case), we would like to propose a contribution, which potentially can improve the memory footprint of the exporters in certain use cases (potentially making it more scalable)

**Proposal:**

Create an extra sampled logger in the telemetry configuration used to avoid flooding the logs with messages that are repeated frequently

The sampled logger is configured by `LogsSamplingConfig`
```
service:
    telemetry:
        logs:
            sampling:
```

Default configuration for the sampled logger 

```
service:
    telemetry:
        logs:
            sampling:
	        initial: 1
	        thereafter: 100
	        tick: 10s
```

### CODE REVIEW

File: exporter/exporterhelper/README.md
Comment (xvilaca-te) [inline exporter/exporterhelper/README.md:27]:
```suggestion
- `sampled_logger`: Enables/disables sampling of logging messages, which caps the CPU and I/O load of logging while keeping a representative subset of your logs.
```

File: exporter/exporterhelper/common.go
Comment (xvilaca-te) [inline exporter/exporterhelper/common.go:34]:
```suggestion
	// Enable/disable the sampledLogger
```

File: exporter/exporterhelper/common.go
Comment (xvilaca-te) [inline exporter/exporterhelper/common.go:30]:
```suggestion
// SampledLoggerSettings configure the logging sampler, which caps the CPU and I/O load of logging while keeping a
```

File: exporter/exporterhelper/queued_retry.go
Comment (xvilaca-te) [inline exporter/exporterhelper/queued_retry.go:275]: Should we merge this with the above if?

File: exporter/exporterhelper/queued_retry_test.go
Comment (xvilaca-te) [inline exporter/exporterhelper/queued_retry_test.go:624]: This is kind of our way of writing parametrised tests. The test cases above seem simpler. Not sure if they will want a test function per test case with a name using snake case.

File: exporter/otlphttpexporter/README.md
Comment (xvilaca-te) [inline exporter/otlphttpexporter/README.md:34]: Same as above.

File: exporter/exporterhelper/queued_retry_test.go
Comment (antonjim-te) [reply exporter/exporterhelper/queued_retry_test.go:624]: They have a similar test in this file https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry_test.go#L389
```
func TestGetRetrySettings(t *testing.T) {
	getStorageClientError := errors.New("unable to create storage client")
	testCases := []struct {
		desc           string
		storage        storage.Extension
		numStorages    int
		storageIndex   int
		expectedError  error
		getClientError error
	}{
		{
			desc:          "obtain storage extension by name",
			numStorages:   2,
			storageIndex:  0,
			expectedError: nil,
		},
		{
			desc:          "fail on not existing storage extension",
			numStorages:   2,
			storageIndex:  100,
			expectedError: errNoStorageClient,
		},
		{
			desc:          "invalid extension type",
			numStorages:   2,
			storageIndex:  100,
			expectedError: errNoStorageClient,
		},
		{
			desc:           "fail on error getting storage client from extension",
			numStorages:    1,
			storageIndex:   0,
			expectedError:  getStorageClientError,
			getClientError: getStorageClientError,
		},
	}

	for _, tC := range testCases {
		t.Run(tC.desc, func(t *testing.T) {
			storageID := component.NewIDWithName("file_storage", strconv.Itoa(tC.storageIndex))

			var extensions = map[component.ID]component.Component{}
			for i := 0; i < tC.numStorages; i++ {
				extensions[component.NewIDWithName("file_storage", strconv.Itoa(i))] = &mockStorageExtension{GetClientError: tC.getClientError}
			}
			host := &mockHost{ext: extensions}
			ownerID := component.NewID("foo_exporter")

			// execute
			client, err := toStorageClient(context.Background(), storageID, host, ownerID, component.DataTypeTraces)

			// verify
			if tC.expectedError != nil {
				assert.ErrorIs(t, err, tC.expectedError)
				assert.Nil(t, client)
			} else {
				assert.NoError(t, err)
				assert.NotNil(t, client)
			}
		})
	}
}
```

File: exporter/exporterhelper/README.md
Comment (sboychen-te) [inline exporter/exporterhelper/README.md:27]:
```suggestion
- `sampled_logger`: Enables/disables sampling of the log messages.
```

we could also replace `Enables/disables` by `Toggles`

File: exporter/exporterhelper/common.go
Comment (sboychen-te) [inline exporter/exporterhelper/common.go:31]: Is this sentence related to logger sampler? If it is I would remove it, would only focus describing settings

File: exporter/exporterhelper/queued_retry.go
Comment (sboychen-te) [inline exporter/exporterhelper/queued_retry.go:270]: I feel like there is a bit of contradictory behavior, it makes sense in terms of backwards compatibility and the code, but conceptually I am not sure (we might need to document it at least):

"Sampler is enabled and logs are in debug mode"

In this case the code will not enable the sampler, but since we have a very explicit configuration which literally says "I want sampling", not sure if this combination might confuse people

Comment (sboychen-te) [reply exporter/exporterhelper/queued_retry.go:270]: I would at least document this behavior in case we will not be proposing the behavior change which will allow the property to overwrite debug mode setting

File: exporter/otlpexporter/config.go
Comment (sboychen-te) [inline exporter/otlpexporter/config.go:21]: Would move this up, next to 
```
exporterhelper.RetrySettings   `mapstructure:"retry_on_failure"`
```

File: exporter/otlphttpexporter/README.md
Comment (sboychen-te) [inline exporter/otlphttpexporter/README.md:34]:
```suggestion
- `sampled_logger`: Enables/disables sampling of the log messages.
```

File: exporter/otlphttpexporter/config.go
Comment (sboychen-te) [inline exporter/otlphttpexporter/config.go:29]: Would move this up

File: exporter/exporterhelper/README.md
Comment (antonjim-te) [reply exporter/exporterhelper/README.md]: Keeping `Enables/disables ` for the moment

File: exporter/exporterhelper/queued_retry.go
Comment (sboychen-te) [inline exporter/exporterhelper/queued_retry.go:271]:
```suggestion
		// Debugging is enabled or log sampling is disabled. Don't do any sampling.
```

File: exporter/otlpexporter/config.go
Comment (sboychen-te) [inline exporter/otlpexporter/config.go:19]: Better to leave one line space here, like it was before

**Review by sboychen-te — APPROVED** [review]

**Review by xvilaca-te — APPROVED** [review]

**Review by pedro-te — APPROVED** [review]

**Comment by antonjim-te** [comment]
Hi @bogdandrutu, could you take a look at the PR please?
Thank you.

**Comment by antonjim-te** [comment]
Hi @jpkrohling @bogdandrutu @mx-psi @jmacd @dmitryax @codeboten @open-telemetry/collector-approvers,
Please review it whenever you have a chance. It will solve a memory issue that we are facing.
Thank you in advance.

File: exporter/exporterhelper/queued_retry.go
Comment (jpkrohling) [inline exporter/exporterhelper/queued_retry.go:276]: @codeboten, do you know why we create a sampled logger here? Shouldn't loggers be created based on the telemetry configuration alone? Instead of this PR, I would consider removing this sampled logger altogether.

Comment (mx-psi) [reply exporter/exporterhelper/queued_retry.go:276]: This was added in #2020 by @tigrannajaryan. The PR description says

> All logging is heavily sampled to avoid flooding the lo

File: exporter/exporterhelper/README.md
Comment (tigrannajaryan) [inline exporter/exporterhelper/README.md:27]: It is not clear what does enabling/disabling sampling do. Please expand the description.

Comment (antonjim-te) [reply exporter/exporterhelper/README.md]: @tigrannajaryan thank you for your feedback.
Is it now more clear? Am I missing something?

Thanks

File: exporter/exporterhelper/queued_retry.go
Comment (jpkrohling) [reply exporter/exporterhelper/queued_retry.go:276]: Right, and shouldn't this be a collector-wide setting, instead of something exclusively to this helper?

Comment (mx-psi) [reply exporter/exporterhelper/queued_retry.go:276]: I think this was meant to be specific to this helper to avoid flooding the logs in the case where all requests fail, and you log once per request. The [contributing guidelines][1] say

> Use logging carefully for events that can happen frequently to avoid flooding the logs. Avoid outputting logs per a received or processed data item since this can amount to very large number of log entries (Collector is designed to process many thousands of spans and metrics per second). For such high-frequency events instead of logging consider adding an internal metric and increment it when the event happens.

I guess at the time this was written metrics were not really available and thus we went with sampled logs.

(Not sure if this is a good reason today to keep this, just trying to reconstruct the history of this to make a better decision)

[1]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#logging

Comment (jpkrohling) [reply exporter/exporterhelper/queued_retry.go:276]: Same here: I'm sure this was a good decision by then, not so sure the next iteration should be to expose a setting for this instead of using a metric and/or exposing log sampling as part of the collector's telemetry config.

Comment (mx-psi) [reply exporter/exporterhelper/queued_retry.go:276]: @tigrannajaryan any thoughts on the above?

Comment (tigrannajaryan) [reply exporter/exporterhelper/queued_retry.go:276]: I think a possible future option is to supply 2 loggers to every component: one that has no sampling (outputs 100% of logs) and another that is sampled. Components will use the no-sampled logger for everything that is critical to be printed and is known to be non-repeating so it can't flood logs (e.g. startup and shutdown messages. The messages that can repeat for every data item or request will use the sampled logger. The sampled logger can derive its sampling ratio from a central setting (somewhere in the `telemetry` config section).

I do not think merely passing a single sampled logger to every component to be used for every message is good enough since it can result in critical one-off messages being lost.

On a side note: the memory usage by sampled logger that this PR attempts to address is puzzling. Maybe try to figure out why and reduce that instead of disabling sampling?

Comment (antonjim-te) [reply exporter/exporterhelper/queued_retry.go:276]: Hi @tigrannajaryan answering the side note. As you can see on the graph https://user-images.githubusercontent.com/5960625/255850931-f9b3e891-2695-4aac-b8d6-9d60cb0dfb9d.png the memory problem is coming from the logger sampling. We also tried modifying the configuration of `NewSamplerWithOptions` 
```
zapcore.NewSamplerWithOptions(
			core,
			10*time.Second,
			1,
			100,
		)
```
But it was still a high memory consumption.
So, it would be great to disable the logger sampler somehow from the `telemetry` config (in a long-term solution) or from the `exporterHelper` config (short-term solution)

Thank you.

Comment (dmitryax) [reply exporter/exporterhelper/queued_retry.go:276]: I agree with @tigrannajaryan. This doesn't seem like a good option to expose in the exporters' configuration interface.

@antonjim-te can you please try initializing one sampled logger instance per service and see if that helps with memory utilization? It can be created in `service.New` function and passed as another field of `TelemetrySettings`  to all components. If it doesn't help with memory utilization, we can still expose a similar configuration for the sampled logger in the `service::telemetry`.

Comment (antonjim-te) [reply exporter/exporterhelper/queued_retry.go:276]: Hi @dmitryax ,
Thank you for your suggestion. I will work on that in a separate branch, test it and if it solves the problem I will bring those changes to into this PR.

I will create a ticket in our internal process and add it to our next sprint.

I have a good feeling about this suggestion because it will just create one logger sampled for all the exporters instead of creating one per `exporterHelper`.

Thank you, team. I will keep you updated.

Comment (antonjim-te) [reply exporter/exporterhelper/queued_retry.go]: Hi team, 
I have followed the suggestion from @dmitryax and created an extra sampled logger in the `TelemetrySettings` and used it when necessary in our case in the `QueuedRetrySender`

Please take a look at the new PR changes.

Thank you in advance.

File: component/telemetry.go
Comment (mx-psi) [inline component/telemetry.go:22]: `SampledLogger` should always be built from `Logger`. Instead of doing it like this, if we want to have this as part of the public API, I think it is better to have a method on `TelemetrySettings` that generates the `SampledLogger` on first use (e.g. using `sync.Once` to initialize a private field).

Alternatively, I would question that we need this on `TelemetrySettings` (do we have any other use cases for it?). Can we do it so that this is internal to exporterhelper? Without changing the public API, the way to do that would be a private global variable initialized on first use.

Comment (antonjim-te) [reply component/telemetry.go]: Great feedback. I will create the sampledLogger the first time using it

**Comment by antonjim-te** [comment]
Hi @mx-psi , @dmitryax , @bogdandrutu  , @tigrannajaryan could you please take a look at the PR with the lastest feedback applied.
Thank you.

File: service/telemetry/telemetry.go
Comment (dmitryax) [inline service/telemetry/telemetry.go:62]: In this current form, this is a breaking change that drops existing configuration functionality. 

`LogsSamplingConfig` was applied to all the logs. Now, we are changing it to be applied to exporter logs only. There are other places where the logs can be pretty noisy when triggered by data handling errors. This config option was initially introduced to resolve that specific issue in the batch processor: https://github.com/open-telemetry/opentelemetry-collector/issues/4554. And now, we are dropping that functionality.

Until we pass the sampledLogger to all the possibly noisy outputs, we cannot apply this change. But I don't think it's necessary at all.

I suggest we just introduce a reasonable default to the existing sampling configuration and remove the additional hardcoded sampling in the exporterhelper. By default, we can probably sample 100 logs in a second after 10 of the initial logs. That should be enough not to miss important logs and ensure it's applied to the noisy ones only. This will simplify the code and will not introduce any additional configuration options or exposed Go API.

@open-telemetry/collector-approvers WDYT?

Comment (antonjim-te) [reply service/telemetry/telemetry.go]: Hi @dmitryax ,
I agree we should not break customer already using `Sampling` in the general logger. I kept this configuration.

Setting a default configuration for the extra sampled logger.

For future work, we should use this sampled logger for noisy outputs. 

Thank you.

**Review by jmacd — APPROVED** [review]

File: service/telemetry/telemetry.go
Comment (dmitryax) [reply service/telemetry/telemetry.go]: Discussed the suggested approach during the Collector SIG meeting. No objections to going ahead with it. We need to make sure it's documented with instructions on how to disable the sampling

File: service/telemetry/telemetry.go
Comment (dmitryax) [inline service/telemetry/telemetry.go:23]: Based on the recent discussion, we don't need any additional sampling logger. We just need to introduce the default sampling configuration for the `logger`

Comment (antonjim-te) [reply service/telemetry/telemetry.go]: Hi @dmitryax ,
Thank you for reviewing this topic in the SIG meeting.

Default configuration
```
					Sampling: &telemetry.LogsSamplingConfig{
						Enabled:    true,
						Tick:       1 * time.Second,
						Initial:    10,
						Thereafter: 100,
					},
```

It can be disabled easily by setting `enabled` to `true`

Thanks

File: otelcol/unmarshaler.go
Comment (dmitryax) [inline otelcol/unmarshaler.go:51]: I think let's increase the tick and make it 10 seconds. It'll be more aligned with the current configuration on the exporter side. Otherwise, after this change, it will give us 100x more logs.

File: service/telemetry/telemetry_test.go
Comment (dmitryax) [inline service/telemetry/telemetry_test.go:18]: Is this test related to the change?

File: service/telemetry/telemetry.go
Comment (dmitryax) [inline service/telemetry/telemetry.go:89]: The comment is a bit confusing. Should be something like this

```suggestion
	// Create a logger that samples every Nth message after the first M messages every S seconds 
	// where N = sc.Thereafter, M = sc.Initial, S = sc.Tick.
```

File: service/telemetry/config.go
Comment (dmitryax) [inline service/telemetry/config.go:105]: Can you please add comment to the each field?

**Review by dmitryax — COMMENTED** [review]
Thank you, @antonjim-te. It looks good to me. Just minor comment

File: service/telemetry/telemetry_test.go
Comment (antonjim-te) [reply service/telemetry/telemetry_test.go:18]: No, it was more in the side of testing the validation of the Telemetry Configuration

File: service/telemetry/config.go
Comment (dmitryax) [inline service/telemetry/config.go:108]:
```suggestion
	// Thereafter represents the sampling rate, every Nth message will be sampled after Initial messages are logged during each Tick.
```

File: .chloggen/SampledLoggerTelemetry.yaml
Comment (dmitryax) [inline .chloggen/SampledLoggerTelemetry.yaml:18]:
```suggestion
subtext: The sampled logger configuration can be disabled easily by setting the `service::telemetry::logs::sampling::enabled` to `false`.
```

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
I'd like to have another look from any of @open-telemetry/collector-approvers before merging

File: otelcol/unmarshaler.go
Comment (antonjim-te) [reply otelcol/unmarshaler.go:52]: Agree 
Rollback the initial to 100 as it as initially

Comment (dmitryax) [reply otelcol/unmarshaler.go]: I don't think that's too low. Currently exporters are sampling from the 1 log. So changing is to 10 makes the exporters already 10x chatty than they used to be. I think making them 100x chatty is too much

Comment (antonjim-te) [reply otelcol/unmarshaler.go]: I see your point. 
@jpkrohling @dmitryax What should be the default? Should the `Initial` be 20?
```
					Sampling: &telemetry.LogsSamplingConfig{
						Enabled:    true,
						Tick:       10 * time.Second,
						Initial:    20,
						Thereafter: 100,
					},
```
Thanks

Comment (jpkrohling) [reply otelcol/unmarshaler.go]: I guess the docs are confusing to me: is it about logging the _same_ messages 10 times, or _any_ messages 10 times? If it's about skipping the 11th _repeated_ message within any 10s window, I agree that 10 is a good number.

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
